### PR TITLE
8339133: [8u] Profiler crashes at guarantee(is_result_safe || is_in_asgct()): unsafe access to zombie method

### PR DESCRIPTION
**Pull request by apangin**
When running AsyncGetCallTrace based profiler, JVM occasionally crashes on the guarantee at codeCache.cpp:287:
```
guarantee(is_result_safe || is_in_asgct(), "unsafe access to zombie method");
```
According to the stack trace, a thread is inside `AsyncGetCallTrace` at the moment of crash, but `is_in_asgct() == false`.

Current implementation of `AsyncGetCallTrace` is not reentrant, since it incorrectly resets `in_asgct` flag of the current thread. The fix is similar to [JDK-8329103](https://bugs.openjdk.org/browse/JDK-8329103): `in_asgct` should be reset to previous value instead of `false`.

The fix is trivial and safe: it affects only `AsyncGetCallTrace`.

Testing: tier1, renaissance suite with async-profiler enabled

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] [JDK-8339133](https://bugs.openjdk.org/browse/JDK-8339133) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339133](https://bugs.openjdk.org/browse/JDK-8339133): [8u] Profiler crashes at guarantee(is_result_safe || is_in_asgct()): unsafe access to zombie method (**Bug** - P4 - Approved)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/568/head:pull/568` \
`$ git checkout pull/568`

Update a local copy of the PR: \
`$ git checkout pull/568` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/568/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 568`

View PR using the GUI difftool: \
`$ git pr show -t 568`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/568.diff">https://git.openjdk.org/jdk8u-dev/pull/568.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/568#issuecomment-2313959218)

### CODE REVIEW

**Comment by tstuefe** [comment]
Hi Andrey, why do you submit this as a jdk 8 PR? Why not fix it upstream? Is it not an issue on JDK 11/17/21 etc?

**Comment by apangin** [comment]
@tstuefe The issue has been already fixed in upstream in JDK-8329103 and then backported to 21u, 17u and 11u.
I created a separate issue for 8u instead of calling it a backport because: (1) in jdk8u, there is no assertion mentioned in the title of JDK-8329103; (2) the patch for JDK-8329103 does not apply to jdk8u, the fix needs to be implemented differently, although the idea is essentially the same.

**Comment by tstuefe** [comment]
> @tstuefe The issue has been already fixed in upstream in JDK-8329103 and then backported to 21u, 17u and 11u. I created a separate issue for 8u instead of calling it a backport because: (1) in jdk8u, there is no assertion mentioned in the title of JDK-8329103; (2) the patch for JDK-8329103 does not apply to jdk8u, the fix needs to be implemented differently, although the idea is essentially the same.

Okay, thank you for the explanation. Up to @gnu-andrew to decide if this is the right approach. Technically, the patch is fine.

**Review by tstuefe — APPROVED** [review]
LGTM

**Comment by apangin** [comment]
/approval request Low risk yet important fix for profiling robustness. Affects only AsyncGetCallTrace code.

**Comment by apangin** [comment]
Dear maintainers, is there anything else I need to do to move this PR forward?

**Comment by tstuefe** [comment]
Ping @jerboaa or @gnu-andrew . Patch looks fine to me.

**Review by shipilev — APPROVED** [review]

**Comment by shipilev** [comment]
I looked at GHA failures, and they do not look relevant to this PR.

/sponsor

**Pull request closed without merging**
